composer/core/time.py

```diff
@@ -0,0 +1,505 @@
+        assert self.value > other.value
+        return 1
+
+    def __eq__(self, other: object):
+        return self._cmp(other) == 0
+
+    def __ne__(self, other: object):
+        return self._cmp(other) != 0
+
+    def __lt__(self, other: object):
+        return self._cmp(other) < 0
+
+    def __le__(self, other: object):
+        return self._cmp(other) <= 0
+
+    def __gt__(self, other: object):
+        return self._cmp(other) > 0
+
+    def __ge__(self, other: object):
+        return self._cmp(other) >= 0
+
+    def __add__(self, other: object):
+        other = self._convert(other)
+        if self.unit != other.unit:
+            raise RuntimeError(f"Cannot add {self} to {other} since they have different units.")
+        return Time(self.value + other.value, self.unit)
+
+    def __radd__(self, other: object):
+        return self + other
+
+    def __sub__(self, other: object):
+        other = self._convert(other)
+        if self.unit != other.unit:
+            raise RuntimeError(f"Cannot subtract {other} from {self} since they have different units.")
+        return Time(self.value - other.value, self.unit)
+
+    def __rsub__(self, other: object):
+        return (-self) + other
+
+    def __neg__(self):
+        return Time(-self.value, self.unit)
+
+    def __pos__(self):
+        return Time(self)
+
+    def __int__(self):
+        return int(self.value)
+
+    def __float__(self):
+        return float(self.value)
+
+    def __truediv__(self, other: object):
+        other = self._convert(other)
+        if self.unit != other.unit:
+            raise RuntimeError(f"Cannot divide {self} by {other} since they have different units.")
+        return Time(self.value / other.value, TimeUnit.DURATION)
+
+    def __mul__(self, other: object):
+        other = self._convert(other)
+        if other.unit != TimeUnit.DURATION and self.unit != TimeUnit.DURATION:
+            raise RuntimeError(f"Multiplication is supported only if one of the units is Duration")
+        real_unit = self.unit if other.unit == TimeUnit.DURATION else other.unit
+        return Time(self.value * other.value, real_unit)
+
+    def __rmul__(self, other: object):
+        return self * other
+
+    @staticmethod
+    def _parse_time_string(timestring: str) -> Tuple[Union[int, float], TimeUnit]:
+        # Parses a time string into a (value, TimeUnit) tuple
+        match = _TIME_STR_REGEX.findall(timestring)
+        if len(match) != 1:
+            raise ValueError(f"Invalid time string: {timestring}")
+        match = match[0]
+        match = [x for x in match if x != '']
+        assert len(match) == 2, "each match should have a number followed by the key"
+        value = match[0]
+        unit = TimeUnit(match[1])
+        value = float(value)  # always parsing first as float b/c it could be scientific notation
+        if unit != TimeUnit.DURATION:
+            value = int(value)
+        return value, unit
+
+    def convert(
```

Comment:
I mocked this out for `to_epoch` an mostly every parameter would still have to be optional -- it depends a lot on what the from unit is. I can add in a matrix into the docs to make it clearer.

Reply:
I added in helper functions per @Averylamp's suggestion

Reply:
So, looking back at #146 and the ways we want `Time` and `Timer` to be used, I think `Time` conversions should be considered super dangerous, and should maybe even log a warning. They cannot be guaranteed to be correct either forwards or backwards in time, because we could be dynamically editing batch size, seq_len, filtering out data from the dataset, etc. 

Ideally, the only object that should ever perform `Time` conversions is the `Trainer` upon initialization, to convert milestones to a shared base `TimeUnit` and even then I'm not sure if it is actually needed... see this example YAML (https://github.com/mosaicml/composer/issues/146#issuecomment-994031037). The key question is "how do I specify that all durations relative to `90ep` should be measured in samples `sp`? Because I need the warmup period to be 5% of `90ep`, but granular down to the sample."

To this end, I would prefer if `Time` conversion was only something that a `Timer` or `Trainer` object could do but not attached to `Time`.

One related issue with naming, the names like `dataset_num_samples`, `batch_size` is all very `Trainer` specific, what is really needed for conversion is `num_samples_per_epoch`, `samples_per_batch`, `tokens_per_batch`... The former is sort of naming what `Trainer` would pass as the values to the convert function, which requires knowing how dataloaders + DDP work

Reply:
One spec I can imagine is, `Time` units have no conversions, and the `Timer` is given a `base_unit` which tells it the granularity with which to increment its internal `timer._duration`. This way, any milestones specified like `0.1dur` will be able to compare with `Timer`'s duration and figure out when to trigger.

Also, `Timer` may need to have both global and period counters, since we will likely want to say things like "`0.1dur` of the post-warmup period".

Reply:
That makes sense -- I don't think users should need to call convert. I can make it private. In that case, since it will be private, is it OK to have the one function with a bunch of optional parameters?

Reply:
Re: duration of training periods, would it be OK if we did this as a separate feature? It would be helpful for me if we had a more detailed design proposal for this (e.g. how would users specify the length of periods? Can you mix and match units for period duration? How many periods would there be? Can periods be user-defined in the yaml?) I don't think it would be too hard to extend the timing object to support a duration relative to a period, but would prefer to get this core functionality merged in first.